File: src/EflayGameSaveManager.Core/Models/ManagerConfig.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace EflayGameSaveManager.Core.Models;

public sealed class ManagerConfig
{
    [JsonPropertyName("version")]
    public string Version { get; set; } = string.Empty;

    [JsonPropertyName("backup_path")]
    public string BackupPath { get; set; } = "./save_data";

    [JsonPropertyName("games")]
    public List<GameDefinition> Games { get; set; } = [];

    [JsonPropertyName("settings")]
    public AppSettings Settings { get; set; } = new();

    [JsonPropertyName("favorites")]
    public List<FavoriteNode> Favorites { get; set; } = [];

    [JsonPropertyName("quick_action")]
    public QuickActionSettings QuickAction { get; set; } = new();

    [JsonPropertyName("devices")]
    public Dictionary<string, DeviceDefinition> Devices { get; set; } = [];
}

public sealed class GameDefinition
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("save_paths")]
    public List<SaveUnitDefinition> SavePaths { get; set; } = [];

    [JsonPropertyName("game_paths")]
    public Dictionary<string, string> GamePaths { get; set; } = [];

    [JsonPropertyName("next_save_unit_id")]
    public int NextSaveUnitId { get; set; }

    [JsonPropertyName("cloud_sync_enabled")]
    public bool CloudSyncEnabled { get; set; }
}

public sealed class SaveUnitDefinition
{
    [JsonPropertyName("id")]
    public int Id { get; set; }

    [JsonPropertyName("unit_type")]
    public SaveUnitType UnitType { get; set; }

    [JsonPropertyName("paths")]
    public Dictionary<string, string> Paths { get; set; } = [];

    [JsonPropertyName("delete_before_apply")]
    public bool DeleteBeforeApply { get; set; }
}

public sealed class AppSettings
{
    [JsonPropertyName("locale")]
    public string Locale { get; set; } = string.Empty;

    [JsonPropertyName("log_to_file")]
    public bool LogToFile { get; set; }

    [JsonPropertyName("cloud_settings")]
    public CloudSettings CloudSettings { get; set; } = new();
}

public sealed class CloudSettings
{
    [JsonPropertyName("always_sync")]
    public bool AlwaysSync { get; set; }

    [JsonPropertyName("auto_sync_interval")]
    public int AutoSyncInterval { get; set; }

    [JsonPropertyName("root_path")]
    public string RootPath { get; set; } = string.Empty;

    [JsonPropertyName("backend")]
    public CloudBackendSettings Backend { get; set; } = new();

    [JsonPropertyName("max_concurrency")]
    public int MaxConcurrency { get; set; }
}

public sealed class CloudBackendSettings
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("endpoint")]
    public string Endpoint { get; set; } = string.Empty;

    [JsonPropertyName("bucket")]
    public string Bucket { get; set; } = string.Empty;

    [JsonPropertyName("region")]
    public string Region { get; set; } = string.Empty;

    [JsonPropertyName("access_key_id")]
    public string AccessKeyId { get; set; } = string.Empty;

    [JsonPropertyName("secret_access_key")]
    public string SecretAccessKey { get; set; } = string.Empty;
}

public sealed class FavoriteNode
{
    [JsonPropertyName("node_id")]
    public string NodeId { get; set; } = string.Empty;

    [JsonPropertyName("label")]
    public string Label { get; set; } = string.Empty;

    [JsonPropertyName("is_leaf")]
    public bool IsLeaf { get; set; }

    [JsonPropertyName("children")]
    public List<FavoriteNode>? Children { get; set; }
}

public sealed class QuickActionSettings
{
    [JsonPropertyName("quick_action_game")]
    public JsonElement QuickActionGame { get; set; }

    [JsonPropertyName("hotkeys")]
    public HotkeySettings Hotkeys { get; set; } = new();
}

public sealed class HotkeySettings
{
    [JsonPropertyName("apply")]
    public List<string> Apply { get; set; } = [];

    [JsonPropertyName("backup")]
    public List<string> Backup { get; set; } = [];
}

public sealed class DeviceDefinition
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;
}

File: src/EflayGameSaveManager.Kotlin/app/src/main/java/com/eflay/gamesavemanager/model/ManagerConfig.kt
package com.eflay.gamesavemanager.model

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
data class ManagerConfig(
    val version: String = "",
    val backup_path: String = "./save_data",
    val games: List<GameDefinition> = emptyList(),
    val settings: AppSettings = AppSettings(),
    val favorites: List<FavoriteNode> = emptyList(),
    val quick_action: QuickActionSettings = QuickActionSettings(),
    val devices: Map<String, DeviceDefinition> = emptyMap()
)

@Serializable
data class GameDefinition(
    val name: String = "",
    val save_paths: List<SaveUnitDefinition> = emptyList(),
    val game_paths: Map<String, String> = emptyMap(),
    val next_save_unit_id: Int = 0,
    val cloud_sync_enabled: Boolean = true
)

@Serializable
data class SaveUnitDefinition(
    val id: Int = 0,
    val unit_type: SaveUnitType = SaveUnitType.Folder,
    val paths: Map<String, String> = emptyMap(),
    val delete_before_apply: Boolean = false,
    val linked_unit_ids: List<Int> = emptyList()
)

@Serializable
enum class SaveUnitType {
    Folder,
    File,
    WinRegistry,
    Zip
}

@Serializable
data class AppSettings(
    val locale: String = "",
    val log_to_file: Boolean = false,
    val cloud_settings: CloudSettings = CloudSettings()
)

@Serializable
data class CloudSettings(
    val always_sync: Boolean = false,
    val auto_sync_interval: Int = 0,
    val root_path: String = "",
    val backend: CloudBackendSettings = CloudBackendSettings(),
    val max_concurrency: Int = 1
)

@Serializable
data class CloudBackendSettings(
    val type: String = "",
    val endpoint: String = "",
    val bucket: String = "",
    val region: String = "",
    val access_key_id: String = "",
    val secret_access_key: String = ""
)

@Serializable
data class FavoriteNode(
    val node_id: String = "",
    val label: String = "",
    val is_leaf: Boolean = true,
    val children: List<FavoriteNode>? = null
)

@Serializable
data class QuickActionSettings(
    val quick_action_game: JsonElement? = null,
    val hotkeys: HotkeySettings = HotkeySettings()
)

@Serializable
data class HotkeySettings(
    val apply: List<String> = emptyList(),
    val backup: List<String> = emptyList()
)

@Serializable
data class DeviceDefinition(
    val id: String = "",
    val name: String = ""
)
